v2/snmp/types.go

```diff
@@ -0,0 +1,151 @@
+	resolvedTimeTag      = timeTag & tagMask
+	opaqueTag            = 0x44
+	resolvedOpaqueTag    = opaqueTag & tagMask
+	counter64Tag         = 0x46
+	resolvedCounter64Tag = counter64Tag & tagMask
+
+	endOfMibTag               = 0x82
+	resolvedEndOfMibTag       = endOfMibTag & tagMask
+	noSuchObjectTag           = 0x80
+	resolvedNoSuchObjectTag   = noSuchObjectTag & tagMask
+	noSuchInstanceTag         = 0x81
+	resolvedNoSuchInstanceTag = noSuchInstanceTag & tagMask
+)
+
+// DataType is used to define the different types of variable found in variable bindings.
+type DataType int
+
+const (
+	Integer DataType = iota
+	OctetString
+	OID
+
+	IpAdddress
+	Time
+	Counter32
+	Counter64
+	Gauge32
+	Opaque
+
+	EndOfMib
+	NoSuchObject
+	NoSuchInstance
+)
+
+// Unmarshals an asn1 RawValue contqining a single variable to deliver a TypedValue that encapsulates the variable type and
+// the golang representation of the variable value.
+func unmarshalVariable(raw *asn1.RawValue) (*TypedValue, error) {
+
+	switch raw.Class {
+	case asn1.ClassUniversal:
+		switch raw.Tag {
+		case asn1.TagInteger:
+			return unmarshalInteger(raw, Integer)
+		case asn1.TagOctetString:
+			return unmarshalString(raw, OctetString)
+		case asn1.TagOID:
+			return unmarshalOID(raw)
+		}
+
+	case asn1.ClassApplication:
+		switch raw.Tag {
+		case resolvedIpTag:
+			return unmarshalString(raw, IpAdddress)
+		case resolvedCounter32Tag:
+			return unmarshalInteger(raw, Counter32)
+		case resolvedCounter64Tag:
+			return unmarshalInteger(raw, Counter64)
+		case resolvedGauge32Tag:
+			return unmarshalInteger(raw, Gauge32)
+		case resolvedTimeTag:
+			return unmarshalInteger(raw, Time)
+		case resolvedOpaqueTag:
+			return unmarshalString(raw, Opaque)
+		}
+	case asn1.ClassContextSpecific:
+		switch raw.Tag {
+		case resolvedEndOfMibTag:
+			return &TypedValue{Type: EndOfMib}, nil
+		case resolvedNoSuchInstanceTag:
+			return &TypedValue{Type: NoSuchInstance}, nil
+		case resolvedNoSuchObjectTag:
+			return &TypedValue{Type: NoSuchObject}, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unsupported class %d tag %d", raw.Class, raw.Tag)
+}
+
+// Unmarshals an SNMP integer-based variable into a TypedValue.
+func unmarshalInteger(raw *asn1.RawValue, dataType DataType) (*TypedValue, error) {
+	var value int64
+	// Replace SNMP-tag with the generic Integer tag, so ASN1 unmarshalling works.
+	raw.FullBytes[0] = asn1.TagInteger
+	_, err := ber.Unmarshal(raw.FullBytes, &value)
+	if err != nil {
+		return nil, err
+	}
+	return &TypedValue{Type: dataType, Value: integerValue(value, dataType)}, nil
+}
+
+// Casts an integer value to the integer type that corresponds to the SNMP data type.
+func integerValue(v int64, dataType DataType) interface{} {
+	switch dataType {
+	case Counter32:
+		fallthrough
+	case Gauge32:
+		fallthrough
+	case Time:
+		return uint32(v)
+
+	case Counter64:
+		return uint64(v)
+	}
+	return v
+}
+
+// Unmarshals an SNMP octetstring-based variable into a TypedValue.
```

Comment:
should the comment be Unmarshals an String octetstring-based or the function name changed to unmarshalSNMP

Reply:
Will rename.  
Also taking opportunity to add a couple of methods to deliver abstract TypedValue as string, int etc.